pkg/gossip/state/stateupdate.go

```diff
@@ -25,18 +39,64 @@ type ccEventMgrProvider interface {
 	GetMgr() ccapi.EventMgr
 }
 
-// UpdateHandler handles updates to LSCC state
+type appDataHandlerRegistry interface {
+	Register(dataType string, handler appdata.Handler) error
+}
+
+type stateRetriever interface {
+	Retrieve(ctxt context.Context, request *appdata.Request, responseHandler appdata.ResponseHandler, allSet appdata.AllSet, opts ...appdata.Option) (extcommon.Values, error)
+}
+
+type stateDBProvider interface {
+	StateDBForChannel(channelID string) extstatedb.StateDB
+}
+
+type mspProvider interface {
+	MSPID() string
+}
+
+type providers struct {
+	BPProvider      blockPublisherProvider
+	MgrProvider     ccEventMgrProvider
+	MspProvider     mspProvider
+	GossipProvider  common.GossipProvider
+	StateDBProvider stateDBProvider
+	HandlerRegistry appDataHandlerRegistry
+}
+
+// UpdateHandler handles updates to LSCC state and the state cache.
+// As blocks are committed, the state cache is updated on the committing peer and the cache-updates are temporarily stored.
+// When an endorsing peer receives a validated block from the committing peer, a request is made to the committing peer for
+// the cache-updates. When the cache-updates are received then the cache on the endorsing peer is updated (pre-populated).
 type UpdateHandler struct {
-	bpProvider  blockPublisherProvider
-	mgrProvider ccEventMgrProvider
+	*providers
+	cacheUpdaters gcache.Cache
 }
 
-// NewUpdateHandler returns a new state update handler
-func NewUpdateHandler(bpProvider blockPublisherProvider, mgrProvider ccEventMgrProvider) *UpdateHandler {
-	return &UpdateHandler{
-		bpProvider:  bpProvider,
-		mgrProvider: mgrProvider,
+// NewUpdateHandler returns a new state update handler.
+func NewUpdateHandler(providers *providers) *UpdateHandler {
+	logger.Info("Creating new cache update updateHandler")
+
+	h := &UpdateHandler{
+		providers: providers,
+	}
+
+	h.cacheUpdaters = gcache.New(0).LoaderFunc(func(channelID interface{}) (interface{}, error) {
+		return h.createCacheUpdater(channelID.(string)), nil
+	}).Build()
+
+	if roles.IsCommitter() {
+		logger.Info("Registering cache updates request handler")
+
+		if err := h.HandlerRegistry.Register(cacheUpdatesDataType, h.handleCacheUpdatesRequest); err != nil {
+			// Should never happen
+			panic(err)
```

Comment:
just making sure that panic is intentional

Reply:
An error occurs only if you attempt to register the same handler twice. This code is only invoked on startup, so it would definitely be a coding bug if this happened and the panic would immediately point out the bug.